realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -413,38 +414,47 @@ private static void initializeRealm(Realm realm) {
         // the Realm is initialized.
         boolean commitChanges = false;
         try {
-            realm.beginTransaction();
+            // We need to start a transaction no matter readOnly mode, because it acts as an interprocess lock.
+            // TODO: For proper inter-process support we also need to move e.g copying the asset file under an
+            // interprocess lock. This lock can obviously not be created by a Realm instance so we probably need
+            // to implement it in Object Store. When this happens, the `beginTransaction(true)` can be removed again.
+            realm.beginTransaction(true);
```

Comment:
I feel it is not right to deal with the read-only in this case. If it is read-only, we should just open the Realm and validate the schema without involving any write transaction. A read lock will be held on the version when the realm is opened. Multi-processes is not something we can help anyway since, after Realm open in process A, process B could still start a write transaction and change the schema. The schema change handling for read-only will just be the same for read-only realm (well, we don't handle it well now.).